### PR TITLE
Issue #2472263 by vasike, smccabe, mglaman: Create a Shopping cart block

### PR DESCRIPTION
**Pull request by mglaman**

### CODE REVIEW

File: modules/cart/src/Plugin/Block/CartBlock.php [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
+use Drupal\Core\Url;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Provides a Shopping cart block.
+ *
+ * @Block(
+ *   id = "cart",
```
Comment (bojanz) [inline modules/cart/src/Plugin/Block/CartBlock.php]: We should make this "commerce_cart" (namespace it).

File: modules/cart/js/commerce_cart.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+(function ($, Drupal, drupalSettings) {
+  'use strict';
+
+  Drupal.behaviors.commerceCartBlock = {
+    attach: function (context) {
+      var $context = $(context);
+      var $cart = $context.find('.cart--cart-block');
+
+      $cart.hover(
```
Comment (bojanz) [inline modules/cart/js/commerce_cart.js]: We said we'd want it to work on click, not on hover?

File: modules/cart/src/Plugin/Block/CartBlock.php [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+class CartBlock extends BlockBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * The cart provider.
+   *
+   * @var \Drupal\commerce_cart\CartProvider
+   */
+  protected $cartProvider;
+
+
```
Comment (bojanz) [inline modules/cart/src/Plugin/Block/CartBlock.php]: Double newline.

File: modules/cart/src/Plugin/Block/CartBlock.php [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+ *   admin_label = @Translation("Shopping Cart"),
+ *   category = @Translation("Commerce")
+ * )
+ */
+class CartBlock extends BlockBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * The cart provider.
+   *
+   * @var \Drupal\commerce_cart\CartProvider
```
Comment (bojanz) [inline modules/cart/src/Plugin/Block/CartBlock.php]: CartProviderInterface

File: modules/cart/src/Plugin/Block/CartBlock.php [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The Drupal account to use for checking for access to block.
```
Comment (bojanz) [inline modules/cart/src/Plugin/Block/CartBlock.php]: The $account property doesn't seem to be used?

File: modules/cart/src/Plugin/Block/CartBlock.php [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin ID for the plugin instance.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\commerce_cart\CartProviderInterface $cart_provider
+   *   The Cart Provider.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The Entity Type Manager.
```
Comment (bojanz) [inline modules/cart/src/Plugin/Block/CartBlock.php]: "The entity type manager." -> odd casing in the original.

File: modules/cart/src/Plugin/Block/CartBlock.php [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+   * Constructs a new BlockContentBlock.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin ID for the plugin instance.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\commerce_cart\CartProviderInterface $cart_provider
+   *   The Cart Provider.
```
Comment (bojanz) [inline modules/cart/src/Plugin/Block/CartBlock.php]: "The cart provider."

File: modules/cart/src/Plugin/Block/CartBlock.php [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+   * {@inheritdoc}
+   */
+  public function blockForm($form, FormStateInterface $form_state) {
+    $form['cart_item_text'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Items text'),
+      '#default_value' => $this->configuration['item_text'],
+      '#description' => $this->t('This text displays after the product count. Defaults to "items"'),
+    ];
+    $form['cart_cart_link'] = [
```
Comment (bojanz) [inline modules/cart/src/Plugin/Block/CartBlock.php]: "cart_cart_link" is a mistake :)
Sounds like that should not be configurable? Instead, we'd want to have a setting for turning on/off the dropdown?

Comment (mglaman) [inline modules/cart/src/Plugin/Block/CartBlock.php]: Ah, true. Was thinking of the "toggle checkout" link, added this in.

File: modules/cart/js/commerce_cart.js (hunk 2 above)
Comment (mglaman) [inline modules/cart/js/commerce_cart.js]: So I went browsing and it seemed everyone used hover. Clicking is counterintuitive in my opinion. I have to click cart, to open a dropdown, and then click a different cart link in order to view my cart. This supports clicking and going strait to cart

File: modules/cart/src/Plugin/Block/CartBlock.php [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * Constructs a new BlockContentBlock.
```
Comment (bojanz) [inline modules/cart/src/Plugin/Block/CartBlock.php]: "Constructs a new CartBlock object."

**Pull request merged**